### PR TITLE
[Spark] Add support for V2 Checkpoint metadata cleanup

### PR DESCRIPTION
**Pull request by dhruvarya-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Adds metadata cleanup logic for V2 Checkpoints (described in https://github.com/delta-io/delta/issues/1793). MetadataCleanup has been updated to make it aware of V2 Checkpoint sidecar files. Sidecar files are only deleted if 1. They are not being used by any non-expired checkpoints AND 2. They have expired as per the log retention period.

## How was this patch tested?

Added Unit Test that checks for sidecar cleanup.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaRetentionSuite.scala [hunk 1]
```diff
@@ -322,4 +323,171 @@ class DeltaRetentionSuite extends QueryTest
...
+    val snapshot = log.update()
+
+    val checkpointVersion = snapshot.logSegment.checkpointProvider.version
+    logInfo(s"snapshot version: ${snapshot.version} checkpoint: $checkpointVersion")
+
+    log.cleanUpExpiredLogs(snapshot)
+  }
+
+  for (v2CheckpointFormat <- V2Checkpoint.Format.ALL_AS_STRINGS)
+  test(s"sidecar file cleanup [v2CheckpointFormat: $v2CheckpointFormat]") {
```
Comment (prakharjain09) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaRetentionSuite.scala:337]: We should be setting the `v2CheckpointFormat` in the test also - currently it is only used in name.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaRetentionSuite.scala [hunk 2]
```diff
@@ -322,4 +323,171 @@ class DeltaRetentionSuite extends QueryTest
...
+      compareVersions(getDeltaVersions(logPath), "delta", 7 to 10)
+
+      assert(
+        getSidecarFiles(log) ===
+          Set(evenCommitSidecarFile_1, oddCommitSidecarFile_1) ++
+          (7 to 10).map(sidecarFiles(_)))
+    }
+  }
+
+  for (v2CheckpointFormat <- Seq("json", "parquet"))
```
Comment (prakharjain09) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaRetentionSuite.scala:439]: Same as above?

File: spark/src/main/scala/org/apache/spark/sql/delta/MetadataCleanup.scala [hunk 3]
```diff
@@ -126,6 +154,172 @@ trait MetadataCleanup extends DeltaLogging {
...
+    metrics.numActiveJsonCheckpointFiles = jsonCheckpointFiles.size
+    val parquetCheckpointsFileIndex =
+      DeltaLogFileIndex(DeltaLogFileIndex.CHECKPOINT_FILE_FORMAT_PARQUET, parquetCheckpointFiles)
+    val jsonCheckpointsFileIndex =
+      DeltaLogFileIndex(DeltaLogFileIndex.CHECKPOINT_FILE_FORMAT_JSON, jsonCheckpointFiles)
+    val identifyActiveSidecarsStartTimeMs = System.currentTimeMillis()
+    metrics.activeCheckpointsListingTimeTakenMs = identifyActiveSidecarsStartTimeMs - startTimeMs
+    val sparkSession = spark
+    // scalastyle:off sparkimplicits
+    import sparkSession.implicits._
```
Comment (ryan-johnson-databricks) [inline spark/src/main/scala/org/apache/spark/sql/delta/MetadataCleanup.scala:244]: Why needed? The error message explains why it's bad, and also how to import Delta-optimized implicits

**Review by ryan-johnson-databricks — APPROVED** [review]
Very nice!

**Review by prakharjain09 — APPROVED** [review]
Thanks for working on this!

**Pull request closed without merging**